en/docs/extended_topics/implement_data_loader.md

```diff
@@ -0,0 +1,461 @@
+
+```c++
+int32_t PersistentInStream::ReadLine(std::string* l);
+int32_t PersistentInStream::ReadFully(char* s, size_t n);
+```
+
+`ReadLine` function read a row of file to `l` object; `ReadFully` function read `n` bytes data to the memory pointed by `s`. Both functions use 0 as return value on success. 
+
+`MiniDataset` complete the process of file to memory buffer. In next section, we will use `MiniParser` to set the buffer's content to Op's output. 
+
+## MiniParser
+
+`MiniPaser` inherits from `Parser`, we just need to override the `Parser` function. 
+
+```c++
+class MiniParser final : public Parser<TensorBuffer> {
+ public:
+  using LoadTargetPtr = std::shared_ptr<TensorBuffer>;
+  using LoadTargetPtrList = std::vector<LoadTargetPtr>;
+
+  void Parse(std::shared_ptr<LoadTargetPtrList> batch_data,
+             user_op::KernelComputeContext* ctx) override {
+    user_op::Tensor* out_tensor = ctx->Tensor4ArgNameAndIndex("out", 0);
+    double* dptr = out_tensor->mut_dptr<double>();
+
+    MultiThreadLoop(batch_data->size(), [&](size_t i) {
+      TensorBuffer* buffer = batch_data->at(i).get();
+      dptr[i*2]= *(buffer->data<double>());
+      dptr[i*2+1]= *(buffer->data<double>()+1);
+    });
+  }
+};
+```
+
+`Parser` includes 2 parameters, where `batch_data` is an encapsulated `vector`. Each element in this container is the data previously read by `MiniDataSet` through `Next` function. The parameter `ctx` enables us to get the information of Op. Here we mainly obtain the output through `ctx` and get the pointer `dptr` to the output buffer. 
+
+Notice that we use macro `MultiThreadLoop` in the procedure of setting `batch_data`'s data to the Op's output `dptr`. `MultiThreadLoop` allows our loop logic to be executed in multiple threads. It accept 2 parameters, the first parameter is the total number of loop; the second parameter is a callback function, its prototype is `void callback(size_t, i)`. OneFlow will create multiple threads, and call the callback function concurrently. The callback function's parameter `i` indicates the serial number of current loop, allowing us to divide data according to `i` and complete the business logic. 
+
+In the above code, we get the `i`th data in buffer through `batch_data->at(i).get()`, and set it to the location of the `i`th row in the output memory area. There are two columns in total. 
+
+## Data Decoder Operator
+
+Data Decoder operator is a normal operator. It accept the output of `DataReader` as its input, output one or multiple Blobs after some operations. 
+
+In  [ofrecord_decoder_ops.cpp](https://github.com/Oneflow-Inc/oneflow/blob/master/oneflow/user/ops/ofrecord_decoder_ops.cpp), we can see various decoders for OFRecord data. 
+
+The data processed by our Mini Dataloader is simple, so what MiniDecoder does is very simple. It just splits two columns data output from `DataReader` into two one-column outputs as `x` and `y`. 
```

Comment:
The data processed by our Mini Dataloader is simple, so the work done by MiniDecoder is also very simple.